_pylibczi/TypedImage.h

```diff
@@ -0,0 +1,146 @@
+       * @return a vector of smart pointers wrapping Images (2D)
+       */
+      ImVec splitChannels(int start_from_) override;
+// TODO Implement set_sort_order() and operator()<
+  };
+
+  template<typename T>
+  inline T& TypedImage<T>::operator[](const std::vector<size_t>& idxs_xy_)
+  {
+      if (idxs_xy_.size()!=m_shape.size())
+          throw ImageAccessUnderspecifiedException(idxs_xy_.size(), m_shape.size(), "from TypedImage.operator[].");
+      size_t idx = calculateIdx(idxs_xy_);
+      return m_array[idx];
+  }
+
+  template<typename T>
+  inline T* TypedImage<T>::getRawPtr(std::vector<size_t> list_)
+  {
+      std::vector<size_t> zeroPadded(0, m_shape.size());
+      std::copy(list_.rbegin(), list_.rend(), zeroPadded.rbegin());
+      return this->operator[](calculateIdx(zeroPadded));
+  }
+  template<typename T>
+  inline void TypedImage<T>::loadImage(const std::shared_ptr<libCZI::IBitmapData>& bitmap_ptr_, size_t channels_)
+  {
+      libCZI::IntSize size = bitmap_ptr_->GetSize();
+      {
+          libCZI::ScopedBitmapLockerP lckScoped{bitmap_ptr_.get()};
+          // WARNING do not compute the end of the array by multiplying stride by height, they are both uint32_t and you'll get an overflow for larger images
+          uint8_t* sEnd = static_cast<uint8_t*>(lckScoped.ptrDataRoi)+lckScoped.size;
+          SourceRange<T> sourceRange(channels_, static_cast<T*>(lckScoped.ptrDataRoi), (T*) (sEnd), lckScoped.stride, size.w);
+          TargetRange<T> targetRange(channels_, size.w, size.h, m_array.get(), m_array.get()+length());
+          for (std::uint32_t h = 0; h<bitmap_ptr_->GetHeight(); ++h) {
+              paired_for_each(sourceRange.strideBegin(h), sourceRange.strideEnd(h), targetRange.strideBegin(h),
+                  [&](std::vector<T*> src_, std::vector<T*> tgt_) {
+                      paired_for_each(src_.begin(), src_.end(), tgt_.begin(), [&](T* s_, T* t_) {
+                          *t_ = *s_;
+                      });
+                  });
+          }
+      }
+  }
+
+  template<typename T>
+  inline Image::ImVec TypedImage<T>::splitChannels(int start_from_)
+  {
+      ImVec ivec;
+      if (m_shape.size()<3)
+          throw ImageSplitChannelException("TypedImage  only has 2 dimensions. No channels to split.", 0);
+      int cStart = 0;
+      // TODO figure out if C can have a nonzero value for a BGR image
+      if (m_planeCoordinates.TryGetPosition(libCZI::DimensionIndex::C, &cStart) && cStart!=0)
+          throw ImageSplitChannelException("attempting to split channels", cStart);
+      for (int i = 0; i<m_shape[0]; i++) {
+          libCZI::CDimCoordinate tmp(m_planeCoordinates);
+          tmp.Set(libCZI::DimensionIndex::C, i+start_from_); // assign the channel from the BGR
+          // TODO should I change the pixel type from a BGRx to a Grayx/3
+          ivec.emplace_back(new TypedImage<T>({m_shape[1], m_shape[2]}, m_pixelType, &tmp, m_xywh, m_mIndex));
+      }
+      return ivec;
+  }
+
```

Comment:
it could be convenient to typedef (or "using") `TypedImage<uint16_t>` to `Uint16Image`, `TypedImage<uint8_t>` to `Uint8Image`, and `TypedImage<float>` to `FloatImage`.  However, I'm not sure how frequently those come up.